### ISSUE TITLE
Suggestion to rewrite as ??= is incorrect

### ISSUE THREAD

**Issue by aybe** [JetBrains/resharper-unity#1818]
In the following code:

```
using UnityEngine;

public class Billboard : MonoBehaviour
{
    public Camera Camera;

    private void Start()
    {
        if (Camera is null)
        {
            Camera = Camera.main;
        }
    }

    private void Update()
    {
        transform.LookAt(Camera.transform.forward);
    }

    private void OnDrawGizmos()
    {
        Debug.DrawRay(transform.position, transform.forward);
    }
}
```

It suggests to rewrite as `Camera ??= Camera.main` which won't work.

**Comment by aybe** [comment]
It turns out that actually the `is` syntax isn't even valid for Unity !

So, Resharper should instead warn you that such syntax won't produce the expected behavior.

**Comment by citizenmatt** [comment]
Can you check the language level for the project, please? Right click on the project in Solution Explorer and select Properties. You should see language level in there. It looks like R# is recommending C# 8 features, even though Unity doesn't (officially) support C# 8 yet.

You can also take a look at the contents of the `.csproj` file, and look for `LangVersion`. If you're using ReSharper, the project file is generated by the Visual Studio Unity integration, and I'm not sure if they set the language version to a specific supported version (e.g. `7.3`) or to a default value such as `latest`.

Also make sure to install the Unity plugin for ReSharper. This will override ReSharper's view of the language level to something that Unity can handle.

**Comment by aybe** [comment]
I have enabled C# 8.0 (Unity 2020 alpha) and have the plugin installed.

**Comment by citizenmatt** [comment]
OK. Looks like we need to suppress the `??=` suggestion. I think I might have misread the comment about `is null`. Do you mean that this doesn't work at all in Unity, or doesn't work as you'd expect? I.e. it checks if the managed object is null, but doesn't call the custom equality operator so doesn't check the status of the native game object?

**Comment by aybe** [comment]
Exactly, the is operator doesn't return the same result as == does. And since this is a keyword whose behavior can't be overriden, it's probably a good idea that R# warns you about the situation. Le 28 août 2020 10:12, Matt Ellis <notifications@github.com> a écrit :
OK. Looks like we need to suppress the ??= suggestion. I think I might have misread the comment about is null. Do you mean that this doesn't work at all in Unity, or doesn't work as you'd expect? I.e. it checks if the managed object is null, but doesn't call the custom equality operator so doesn't check the status of the native game object?
